### ISSUE TITLE
should check as.tidem()

### ISSUE THREAD

**Issue by dankelley** [dankelley/oce#2142]
This came up because a colleague is using `as.tidem()`, and I realized that I have not checked this well against data.  I'll take a look.

**Comment by dankelley** [comment]
Hm.  Testing for Honolulu, I created a file `const.txt` (attached) by a copying operation on the site https:tidesandcurrents.noaa.gov/harcon.html?id=1612340 and then running the R code
```R
library(oce)
options(width=200)
tRef <- as.POSIXct("2023-08-15", tz="UTC")
latitude <- 21+18.2/60 # https://tidesandcurrents.noaa.gov/stationhome.html?id=1612340

# The const.txt file was created by copying from the website for Hawaii, viz.
# <https:tidesandcurrents.noaa.gov/harcon.html?id=1612340>. The column names are
# 'Constituent..', 'Name', 'Amplitude', 'Phase', 'Speed', and 'Description'.
const <- read.delim("const.txt", sep="\t", header=TRUE)
data(tidedata)
known <- sapply(const$Name, \(n) n %in% tidedata$const$name)
if (!all(known))
    stop("following constituents are not understood: ", paste(const$Name[!known], collapse=" "))
```
but I get
```
Loading required package: gsw
Error: following constituents are not understood: LAM2 M1 RHO 2MK3
Execution halted
```

I'll look into this.

[const.txt](https://github.com/dankelley/oce/files/12479434/const.txt)

**Comment by dankelley** [comment]
Note: using `as.tidem()` also indicates the problem of constituent names.
```R
> library(oce)
Loading required package: gsw
> options(width=200)
> tRef <- as.POSIXct("2023-08-15", tz="UTC")
> latitude <- 21+18.2/60 # https://tidesandcurrents.noaa.gov/stationhome.html?id=1612340
>
> # The const.txt file was created by copying from the website for Hawaii, viz.
> # <https:tidesandcurrents.noaa.gov/harcon.html?id=1612340>. The column names are
> # 'Constituent..', 'Name', 'Amplitude', 'Phase', 'Speed', and 'Description'.
> const <- read.delim("const.txt", sep="\t", header=TRUE)
> tm <- as.tidem(tRef=tRef, latitude=latitude, name=const$Name, amplitude=const$Amplitude, phase=const$Phase)
Warning message:
In as.tidem(tRef = tRef, latitude = latitude, name = const$Name,  :
  the following constituents are not handled: 'LAM2', 'M1', 'RHO', '2MK3'
```

**Comment by dankelley** [comment]
I see at https://cran.r-project.org/web/packages/TideHarmonics/vignettes/austides.pdf (a package I didn't know about!) that others have found odd naming at NOAA sites.

So, I see how to fix the "RHO" problem, I suppose.  My guess is that "LAM2" means "LAMBDA2" or "LAMDA2".

<img width="809" alt="Screenshot 2023-08-30 at 4 27 09 PM" src="https://github.com/dankelley/oce/assets/99469/7ba61778-f2ab-47bc-9ac8-3846cf095e8a">

**Comment by dankelley** [comment]
These are the names oce knows.  Here is where a summer student would come in handy ... if some NOAA site says an unrecognized name has a particular frequency, and if oce has a name with the same frequency, then we know the synonym.

I don't really feel like looking through the names.

```

R version 4.3.1 (2023-06-16) -- "Beagle Scouts"
Copyright (C) 2023 The R Foundation for Statistical Computing
Platform: x86_64-apple-darwin20 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> data("tidedata", package="oce")
> (tidedata$const[c("name","freq")])
    name         freq
1     Z0 0.0000000000
2     SA 0.0001140741
3    SSA 0.0002281591
4    MSM 0.0013097808
5     MM 0.0015121518
6    MSF 0.0028219327
7     MF 0.0030500918
8   ALP1 0.0343965699
9    2Q1 0.0357063507
10  SIG1 0.0359087218
11    Q1 0.0372185026
12  RHO1 0.0374208736
13    O1 0.0387306544
14  TAU1 0.0389588136
15  BET1 0.0400404353
16   NO1 0.0402685944
17  CHI1 0.0404709654
18   PI1 0.0414385130
19    P1 0.0415525871
20    S1 0.0416666721
21    K1 0.0417807462
22  PSI1 0.0418948203
23  PHI1 0.0420089053
24  THE1 0.0430905270
25    J1 0.0432928981
26  2PO1 0.0443745198
27   SO1 0.0446026789
28   OO1 0.0448308380
29  UPS1 0.0463429898
30  ST36 0.0733553835
31  2NS2 0.0746651643
32  ST37 0.0748675353
33   ST1 0.0748933234
34   OQ2 0.0759749451
35  EPS2 0.0761773161
36   ST2 0.0764054753
37   ST3 0.0772331498
38    O2 0.0774613089
39   2N2 0.0774870970
40   MU2 0.0776894680
41  SNK2 0.0787710897
42    N2 0.0789992488
43   NU2 0.0792016198
44   ST4 0.0794555670
45   OP2 0.0802832416
46  GAM2 0.0803090296
47    H1 0.0803973266
48    M2 0.0805114007
49    H2 0.0806254748
50  MKS2 0.0807395598
51   ST5 0.0809677189
52   ST6 0.0815930224
53  LDA2 0.0818211815
54    L2 0.0820235525
55  2SK2 0.0831051742
56    T2 0.0832192592
57    S2 0.0833333333
58    R2 0.0834474074
59    K2 0.0835614924
60  MSN2 0.0848454852
61  ETA2 0.0850736443
62   ST7 0.0853018034
63  2SM2 0.0861552660
64  ST38 0.0863576370
65  SKM2 0.0863834251
66  2SN2 0.0876674179
67   NO3 0.1177299033
68   MO3 0.1192420551
69    M3 0.1207671010
70   NK3 0.1207799950
71   SO3 0.1220639878
72   MK3 0.1222921469
73   SP3 0.1248859204
74   SK3 0.1251140796
75   ST8 0.1566887168
76    N4 0.1579984976
77  3MS4 0.1582008687
78  ST39 0.1592824904
79   MN4 0.1595106495
80   ST9 0.1597388086
81  ST40 0.1607946422
82    M4 0.1610228013
83  ST10 0.1612509604
84   SN4 0.1623325821
85   KN4 0.1625607413
86   MS4 0.1638447340
87   MK4 0.1640728931
88   SL4 0.1653568858
89    S4 0.1666666667
90   SK4 0.1668948258
91  MNO5 0.1982413039
92  2MO5 0.1997534558
93  3MP5 0.1999816149
94  MNK5 0.2012913957
95  2MP5 0.2025753884
96  2MK5 0.2028035475
97  MSK5 0.2056254802
98  3KM5 0.2058536393
99  2SK5 0.2084474129
100 ST11 0.2372259056
101 2NM6 0.2385098983
102 ST12 0.2387380574
103 2MN6 0.2400220501
104 ST13 0.2402502093
105 ST41 0.2413060429
106   M6 0.2415342020
107 MSN6 0.2428439828
108 MKN6 0.2430721419
109 ST42 0.2441279756
110 2MS6 0.2443561347
111 2MK6 0.2445842938
112 NSK6 0.2458940746
113 2SM6 0.2471780673
114 MSK6 0.2474062264
115   S6 0.2500000000
116 ST14 0.2787527046
117 ST15 0.2802906445
118   M7 0.2817899023
119 ST16 0.2830867891
120 3MK7 0.2833149482
121 ST17 0.2861368809
122 ST18 0.3190212990
123 3MN8 0.3205334508
124 ST19 0.3207616099
125   M8 0.3220456027
126 ST20 0.3233553835
127 ST21 0.3235835426
128 3MS8 0.3248675353
129 3MK8 0.3250956944
130 ST22 0.3264054753
131 ST23 0.3276894680
132 ST24 0.3279176271
133 ST25 0.3608020452
134 ST26 0.3623141970
135 4MK9 0.3638263489
136 ST27 0.3666482815
137 ST28 0.4010448515
138  M10 0.4025570033
139 ST29 0.4038667841
140 ST30 0.4053789360
141 ST31 0.4069168759
142 ST32 0.4082008687
143 ST33 0.4471596822
144  M12 0.4830684040
145 ST34 0.4858903367
146 ST35 0.4874282766
```

**Comment by dankelley** [comment]
Here are some notes on naming.  I don't know how to find out more about the NOAA constituent frequencies.

NOAA name to oce name

- [ ] LAM2 I think this is probably lambda2
- [ ] M1 this might be as discussed at https://os.copernicus.org/articles/15/431/2019/ but I don’t think oce has it
- [ ] RHO possibly rho1 https://cran.r-project.org/web/packages/TideHarmonics/vignettes/austides.pdf
- [ ] 2MK3 no idea.  oce has 2MK5 and 2MK6 and some other things but it’s a guessing game

**Comment by dankelley** [comment]
I have some more on M1 now from https://flaterco.com/files/xtide/Constituents-20170508.pdf.  The number is the speed in degrees per hour.

<img width="760" alt="Screenshot 2023-08-30 at 4 47 25 PM" src="https://github.com/dankelley/oce/assets/99469/08de8d2d-aeee-4385-9344-bf8c3f7676be">

**Comment by dankelley** [comment]
I took the extrema of the "M" items in the table above and get, from below, that you'd need 5 years of data to distinguish between the two.  (I might be doing this incorrectly, of course ... just finishing up for the day.)

```R
> a<-14.487410
> b<-14.492052
> 1/(1/a-1/b)
[1] 45228.85
> 1/(1/a-1/b) / 24
[1] 1884.535
> 1/(1/a-1/b) / 24 / 365
[1] 5.16311
```

**Comment by bluepost22** [comment]
Hi Dan -- my OCE gives me that message on the same 4 constituents, but keeps going happily.

BTW most of my work has been done on Kahului, station # 1615680.   Honolulu shows the "holes", but it seems the NE facing harbors show the effect more strongly.

Bob Dow

**Comment by bluepost22** [comment]
And here are my actual R lines:

Here are the two calls in my R code:

tidemodel = as.tidem(tref, latloc, harmfinal$Name, harmfinal$Amplitude, harmfinal$Phase, debug = 0)   # there will be a warning on 4 uncovered harmonics

tidetable = predict.tidem(tidemodel,   newdata=datetimevec, tz="UTC") + msloffset

tidetable returned by predict.tidem is just a vector of tide heights corresponding to the datetimevec.  It is not a dataframe.  I then process "tidetable" looking for local max/mins.  Stackexchange is your friend.

I never figured out if predict.tidem would work with local time inputs -- UTC worked, so I never looked back.

"harmfinal$Name" etc above is the NOAA csv after doing any selection on which harmonics will be used.  "latloc" is the local latitude.

predict.tidem wants you to give it a vector of datetimes and it will give you a vector of tide heights back.  I create that vector:

datetimevec = seq(datetimestart, datetimestop, by=vecinterval)  #  sample series for the wave table, vecinterval = 300 seconds

datetimevecloc = with_tz(datetimevec, tz = tzstring)            # sample series in local time  (you'll want package lubridate in order to keep your sanity)

So I give predict.tidem the x-axis points in datetimevec, but will plot using datetimevecloc as the x-axis points.

For the "tref" input, I used this:

tref = mean(c(ISOdate(1983,1,1, tz="UTC"),ISOdate(2001,12,31, tz="UTC")))

I somewhere got the idea this was the observation interval for all USA.  Ok, it's here, at the bottom, "Tidal Data Analysis Periods":

**Comment by dankelley** [comment]
Oh, I see a table in https://tidesandcurrents.noaa.gov/harcon.html?id=1612340 and also a friend pointed out that the tables (like at https://tidesandcurrents.noaa.gov/harcon.html?id=1612340) give frequency, sort of, via 'speed'.  I don't know why I didn't twig to this conversion.

**Comment by dankelley** [comment]
I figured out the name mapping (as below).  I'll code this translation into oce tomorrow morning, or on Friday.

```R
> library(oce)
> # The const.txt file was created by copying from the website for Hawaii, viz.
> # <https:tidesandcurrents.noaa.gov/harcon.html?id=1612340>. The column names are
> # 'Constituent..', 'Name', 'Amplitude', 'Phase', 'Speed', and 'Description'.
> const <- read.delim("const.txt", sep="\t", header=TRUE)
> const$freq <- const$Speed/360
> data(tidedata)
> known <- sapply(const$Name, \(n) n %in% tidedata$const$name)
> unknownNames <- const[!known, ]$Name
> 
> for (unknownName in unknownNames) {
+     freqDiff <- abs(tidedata$const$freq - const$freq[const$Name==unknownName])
+     w <- which.min(freqDiff)
+     cat("NOAA name \"", unknownName, "\" maps to oce name \"",
+         tidedata$const$name[w], "\" (freq error ", freqDiff[w], " cph)\n", sep="")
+ }
NOAA name "LAM2" maps to oce name "LDA2" (freq error 1.833333e-09 cph)
NOAA name "M1" maps to oce name "NO1" (freq error 4.444444e-11 cph)
NOAA name "RHO" maps to oce name "RHO1" (freq error 1.4e-09 cph)
NOAA name "2MK3" maps to oce name "MO3" (freq error 4.555556e-10 cph)
```

**Comment by dankelley** [comment]
Here is a full accounting, including the matching names and non-matching names.  I am concerned that the `SA` frequency does not match to as many digits as the others.  Notice the line
```
  NOAA "SA" matches oce "SA" (0.0001140741; 0.000114079444444444; 4.68495412372761e-05)
```
The first number is the oce freq, the second is the NOAA freq, and the third is the fractional difference between the two (see the code below).  Converting freq to speed (which NOAA reports) we get 0.04106668 (oce) versus 0.0410686 (NOAA).  This second number is as reported in the NOAA table, to the number of digits in that table.  Thus, the NOAA and oce frequencies for SA differ.

I don't know what to make of this discrepancy, but I think it deserves more study.  Some questions to ponder:
1. Have the best estimates of the frequency changed?  
2. Did my copy/paste from the NOAA site make an error (just for this one case)?  
3. Is the NOAA site simply erroneous, compared with their software?  
4. If the NOAA software differs from that of oce (and t-tide in matlab, and Mike Foreman's code) then who is right, or more right?


<details>
```
> options(digits=15)
> library(oce)
> # The const.txt file was created by copying from the website for Hawaii, viz.
> # <https:tidesandcurrents.noaa.gov/harcon.html?id=1612340>. The column names are
> # 'Constituent..', 'Name', 'Amplitude', 'Phase', 'Speed', and 'Description'.
> const <- read.delim("const.txt", sep="\t", header=TRUE)
> const$Freq <- const$Speed/360
> data(tidedata)
> known <- sapply(const$Name, \(n) n %in% tidedata$const$name)
> unknownNames <- const[!known, ]$Name
> 
> df <- function(a, b) # fractional difference
+     (a-b)/(0.5*(a+b))
> 
> n <- length(const$Name)
> worst <- -1
> wworst <- -1
> for (name in const$Name) {
+     fdf <- abs(df(tidedata$const$freq, const$Freq[const$Name==name]))
+     w <- which.min(fdf)
+     if (name %in% unknownNames) {
+         cat("NOAA \"", name, "\" maps to oce \"",
+             tidedata$const$name[w], "\"",
+             " (", tidedata$const$freq[w], "; ", const$Freq[const$Name==name], "; ",
+             fdf[w], ")\n", sep="")
+     } else {
+         cat("  NOAA \"", name, "\" matches oce \"",
+             name, "\"",
+             " (", tidedata$const$freq[w], "; ", const$Freq[const$Name==name], "; ",
+             fdf[w], ")\n", sep="")
+     }
+     if (fdf[w] > worst) {
+         worst <- fdf[w]
+         wworst <- tidedata$const$name[w]
+     }
+ }
  NOAA "M2" matches oce "M2" (0.0805114007; 0.0805114; 8.69442090640656e-09)
  NOAA "S2" matches oce "S2" (0.0833333333; 0.0833333333333333; 3.99999866642695e-10)
  NOAA "N2" matches oce "N2" (0.0789992488; 0.07899925; 1.51900177692566e-08)
  NOAA "K1" matches oce "K1" (0.0417807462; 0.0417807472222222; 2.44663462734092e-08)
  NOAA "M4" matches oce "M4" (0.1610228013; 0.161022805555556; 2.64282790015117e-08)
  NOAA "O1" matches oce "O1" (0.0387306544; 0.0387306527777778; 4.18847107984744e-08)
  NOAA "M6" matches oce "M6" (0.241534202; 0.241534222222222; 8.37240490832793e-08)
  NOAA "MK3" matches oce "MK3" (0.1222921469; 0.122292147222222; 2.63485630720182e-09)
  NOAA "S4" matches oce "S4" (0.1666666667; 0.166666666666667; 2.00000016528074e-10)
  NOAA "MN4" matches oce "MN4" (0.1595106495; 0.159510644444444; 3.16941574338975e-08)
  NOAA "NU2" matches oce "NU2" (0.0792016198; 0.0792016194444444; 4.48924603959762e-09)
  NOAA "S6" matches oce "S6" (0.25; 0.25; 0)
  NOAA "MU2" matches oce "MU2" (0.077689468; 0.0776894666666667; 1.71623436020065e-08)
  NOAA "2N2" matches oce "2N2" (0.077487097; 0.0774870972222222; 2.86786085862456e-09)
  NOAA "OO1" matches oce "OO1" (0.044830838; 0.0448308361111111; 4.21336965723155e-08)
NOAA "LAM2" maps to oce "LDA2" (0.0818211815; 0.0818211833333333; 2.24065854872945e-08)
  NOAA "S1" matches oce "S1" (0.0416666721; 0.0416666666666667; 1.30399991629124e-07)
NOAA "M1" maps to oce "NO1" (0.0402685944; 0.0402685944444444; 1.10369983905805e-09)
  NOAA "J1" matches oce "J1" (0.0432928981; 0.0432928986111111; 1.18058880597372e-08)
  NOAA "MM" matches oce "MM" (0.0015121518; 0.00151215194444444; 9.55224452801393e-08)
  NOAA "SSA" matches oce "SSA" (0.0002281591; 0.000228159166666667; 2.92193723210994e-07)
  NOAA "SA" matches oce "SA" (0.0001140741; 0.000114079444444444; 4.68495412372761e-05)
  NOAA "MSF" matches oce "MSF" (0.0028219327; 0.00282193277777778; 2.75618823474269e-08)
  NOAA "MF" matches oce "MF" (0.0030500918; 0.00305009194444444; 4.73574078216894e-08)
NOAA "RHO" maps to oce "RHO1" (0.0374208736; 0.037420875; 3.74122738729096e-08)
  NOAA "Q1" matches oce "Q1" (0.0372185026; 0.0372185027777778; 4.77659673994523e-09)
  NOAA "T2" matches oce "T2" (0.0832192592; 0.0832192583333333; 1.04142562155454e-08)
  NOAA "R2" matches oce "R2" (0.0834474074; 0.0834474083333333; 1.11846892689854e-08)
  NOAA "2Q1" matches oce "2Q1" (0.0357063507; 0.03570635; 1.96043559631714e-08)
  NOAA "P1" matches oce "P1" (0.0415525871; 0.0415525861111111; 2.37984918669269e-08)
  NOAA "2SM2" matches oce "2SM2" (0.086155266; 0.0861552666666667; 7.73796784069157e-09)
  NOAA "M3" matches oce "M3" (0.120767101; 0.120767111111111; 8.37240490832793e-08)
  NOAA "L2" matches oce "L2" (0.0820235525; 0.0820235527777778; 3.38656127352044e-09)
NOAA "2MK3" maps to oce "MO3" (0.1192420551; 0.119242055555556; 3.82042698140823e-09)
  NOAA "K2" matches oce "K2" (0.0835614924; 0.0835614944444444; 2.44663462734092e-08)
  NOAA "M8" matches oce "M8" (0.3220456027; 0.322045611111111; 2.6117763945721e-08)
  NOAA "MS4" matches oce "MS4" (0.163844734; 0.163844733333333; 4.06889285603809e-09)
> cat(wworst, " has worst fractional freq match (", worst, " or ", worst*86400, " sec per day)\n", sep="")
SA has worst fractional freq match (4.68495412372761e-05 or 4.04780036290065 sec per day)
```

**Comment by dankelley** [comment]
The copy/paste is correct (see snapshot) so I crossed out possibility 2 from the above.
<img width="745" alt="Screenshot 2023-08-31 at 5 55 17 AM" src="https://github.com/dankelley/oce/assets/99469/20493f98-6739-4349-a4eb-ecfff21b1869">

**Comment by bluepost22** [comment]
Why does OCE have its own idea of SA?  Isn't this an input?

Anyway, there are at least two kinds of "year", Tropical and Sidereal.  

A speed of .0410688 deg/hr yields 365 days, 5 hours, 49 mins, and 16 secs, which sounds like a Tropical year.  decimal:  365.242545

A speed of .04106668 deg/hr  yields 365 days, 6 hours, 13 mins, and 51 seconds, which sounds like a Sidereal year.
decimal:  365.259622

**Comment by dankelley** [comment]
I've code in a test, as shown below.  The base level (which I find in the `zoffset` variable) likely has something to do with the reference height for the predictions.  I'm not interested in that value.  However, I am (a bit) interested in the second panel of the graph.  The small jiggles are likely from the number of digits shown in the NOAA webpage.  I don't know what the slower signal is.  It's too small to care about in a practical sense, but it *is* curious, to a *chat comme moi*.

I'll leave this open for a day or so, in case anyone who is involved in these tide issues has comments.

**Plot**
![2142a](https://github.com/dankelley/oce/assets/99469/60b1a657-6f1f-4662-892d-5e66bd4f97a8)



**Code**

```R
# Compare for station 1612340 (Honolulu).
# I got the data, in file 'predictionFile', from
#https://tidesandcurrents.noaa.gov/waterlevels.html?id=1612340&units=metric&bdate=20230830&edate=20230831&timezone=GMT&datum=MLLW&interval=6&action=data
# and the harmonic coefficients, in file "const.txt", from cut/paste of the table at
# https://tidesandcurrents.noaa.gov/harcon.html?unit=0&timezone=0&id=1612340&name=Honolulu&state=HI

predictionFile <- "CO-OPS_1612340_met.csv"
harmonicFile <- "harmonics.tsv"

library(oce)
options(width=200)
latitude <- 21+18.2/60 # https://tidesandcurrents.noaa.gov/stationhome.html?id=1612340

pred <- read.csv(predictionFile)
head(pred)
pred$time <- as.POSIXct(paste(pred$Date, pred$Time..GMT.), tz="UTC")
head(pred)

tRef <- mean(pred$time)
const <- read.delim(harmonicFile, sep="\t", header=TRUE)
head(const)
m <- as.tidem(tRef=tRef, latitude=latitude, name=const$Name, amplitude=const$Amplitude, phase=const$Phase)
if (!interactive())
    png("2142a.png", unit="in", height=5, width=7, res=200, pointsize=9)
layout(matrix(1:2,ncol=1), height=c(0.6,0.4))
plot(pred$time, pred$Predicted..m., type="l", xlab="Time [UTC] in Year 2023",
    xaxs="i", ylim=c(-0.05, 0.8), lwd=4, col="gray", ylab="Water Level [m]")
mtext("Honolulu (station 1612340)", col=4, font=2, adj=1)
eta <- predict(m, pred$time)
zoffset <- mean(pred$Predicted..m. - eta)
lines(pred$time, eta + zoffset, col=2)
rms <- function(x) sqrt(mean(x^2))
error <- rms(pred$Predicted..m. - (eta + zoffset))
legend("topright", lwd=c(4, 1), col=c("gray", 2), bg="white",
    legend=c("NOAA prediction", paste("oce, with", round(zoffset,4), "m added")))
mtext(sprintf("RMS mismatch after offset: %.4f m", error), adj=0)

plot(pred$time, pred$Predicted..m. - (eta + zoffset),
    type="l", xlab="Time [UTC] in Year 2023", xaxs="i", ylab="Mismatch [m]")
abline(h=0, col=4)

if (!interactive())
    dev.off()
```

**Comment by dankelley** [comment]
@bluepost22 (I think that's the right handle for you, Bob Dow), I followed up on the `tRef` question (which might have been in email, as opposed to here on github).  Below, I'm trying to organize my thoughts using section headings.

**Summary**

I don't think we have any way of knowing the `tRef` that was used in the NOAA analysis.  Perhaps they use the middle time between the start of observations and the most recent observation.  Or perhaps they use the last full year of observations.  It could be anything, really.  Perhaps the interval is documented somewhere.  I don't quite have the time to look for such documentation.  My point here (and in the just-previous comment) is just to show that `tRef` matters to the details.  I think your interest (at the moment) is mainly in times of high and low tides, and these seem to be pretty similar in the plot of this comment and the just-previous one, so you ought to be okay with a variety of choices for `tRef`.

**Methods**

I tried doing the above but with a `tRef` as you had suggested at one point.  The fit is now much worse than before, with the RMS error (after adjusting level empirically) being 5.49 cm, whereas it was 0.025 cm with the other `tRef`.

The only change in the code is the definition of `tRef`, from
```R
Ref <- mean(pred$time)
```
to
```R
tRef <- mean(c(ISOdate(1983,1,1, tz="UTC"),ISOdate(2001,12,31, tz="UTC")))
```
which is what you were using, I think.


**Results**

![2142b](https://github.com/dankelley/oce/assets/99469/d524f8b4-3e75-49e6-af5d-d51025a47e02)



**Appendix: Code**

```R
# Compare for station 1612340 (Honolulu).
# I got the data, in file 'predictionFile', from
#https://tidesandcurrents.noaa.gov/waterlevels.html?id=1612340&units=metric&bdate=20230830&edate=20230831&timezone=GMT&datum=MLLW&interval=6&action=data
# and the harmonic coefficients, in file "const.txt", from cut/paste of the table at
# https://tidesandcurrents.noaa.gov/harcon.html?unit=0&timezone=0&id=1612340&name=Honolulu&state=HI

predictionFile <- "CO-OPS_1612340_met.csv"
harmonicFile <- "harmonics.tsv"

library(oce)
options(width=200)
latitude <- 21+18.2/60 # https://tidesandcurrents.noaa.gov/stationhome.html?id=1612340

pred <- read.csv(predictionFile)
head(pred)
pred$time <- as.POSIXct(paste(pred$Date, pred$Time..GMT.), tz="UTC")
head(pred)

#tRef <- mean(pred$time)
tRef <- mean(c(ISOdate(1983,1,1, tz="UTC"),ISOdate(2001,12,31, tz="UTC")))
const <- read.delim(harmonicFile, sep="\t", header=TRUE)
head(const)
m <- as.tidem(tRef=tRef, latitude=latitude, name=const$Name, amplitude=const$Amplitude, phase=const$Phase)
if (!interactive())
    png("2142b.png", unit="in", height=5, width=7, res=200, pointsize=9)
layout(matrix(1:2,ncol=1), height=c(0.6,0.4))
plot(pred$time, pred$Predicted..m., type="l", xlab="Time [UTC] in Year 2023",
    xaxs="i", ylim=c(-0.05, 0.8), lwd=4, col="gray", ylab="Water Level [m]")
mtext("Honolulu (station 1612340)", col=4, font=2, adj=1)
eta <- predict(m, pred$time)
zoffset <- mean(pred$Predicted..m. - eta)
lines(pred$time, eta + zoffset, col=2)
rms <- function(x) sqrt(mean(x^2))
error <- rms(pred$Predicted..m. - (eta + zoffset))
legend("topright", lwd=c(4, 1), col=c("gray", 2), bg="white",
    legend=c("NOAA prediction", paste("oce, with", round(zoffset,4), "m added")))
mtext(sprintf("RMS mismatch after offset: %.4f m", error), adj=0)

plot(pred$time, pred$Predicted..m. - (eta + zoffset),
    type="l", xlab="Time [UTC] in Year 2023", xaxs="i", ylab="Mismatch [m]")
abline(h=0, col=4)

if (!interactive())
    dev.off()
```

**Comment by dankelley** [comment]
I will build in some tests on the weekend, based on the existing examples for the function, along with the tests with the NOAA case in #2146 etc.

**Comment by dankelley** [comment]
Done in commit 819155f53a30d4b0723586c039030d93c27b9506 of the "develop" branch, in `tests/testthat/test_as_tidem.R`.  Note that I only use a portion of the time-series I was using in this issue, because I think that is enough to catch errors, and I wanted to save space.  (The package is large and so I try to use minimal test data.)

Importantly, the test I have on agreement will fail over a one-day period.  That's because of this low-frequency variation you can see in plots above, which is still a mystery to me.

My plan is to leave this issue open until @richardsc and I can chat about it.

**Comment by dankelley** [comment]
I'm not sure on the offset.  I determined this simply by minimizing the mean between the two predictions.  My goal was really to see if we had gross errors (like time being out by a timezone, etc).

I don't have a theory on the deviation.  You can get my actual test (not the one above, which was provisional) by cloning github.com/dankelley/oce-issues and then doing
```
cd oce-issues/21xx/2146
make
```
As you can see in the R file that gets used by that action (https://github.com/dankelley/oce-issues/blob/68b36dd7592ecf995fd1a53882fd52bcea45f2be/21xx/2146/2146a.R#L28) I get almost exactly the same results whether I use the NOAA speed or the built-in oce frequency.  I think that is because there is low sensitivity to the SA frequency, and the other constituents have frequencies that are very, very similar between NOAA and oce.  That close similarity is seen by the output you'll get if you do as above (obviously, after building the very-latest "develop" commit ov oce):
```
In as.tidem(tRef = mean(o$time), latitude = latitude, name = h$Name,  :
  SA: absolute fractional mismatch between specified and built-in frequency is 4.685e-05
```
The code emits a message like that if the relative shift in freq exceeds 1e-6, and that is a very low number.  I've not played with altering the numbers but the fact that the graph is visually identical tells me that this is not the explanation for that low-frequency pattern.

That pattern is about 3X larger than we would get from a rounding error in the NOAA stated predictions.  But the temporal aspect of the pattern suggests this is not a rounding error.

I have no idea what causes that variation.  I've tried a series of ideas but none have panned out.  Maybe 'u' and 'v' are a bit wrong?  That's hard to tell because NOAA does not state them in their webpages.

I propose to close the issue.  Doing that does not mean that I assess the agreement as being perfect.  It just means -- from the issue title -- that I have inserted some checks.  The point of the checks is to catch future code shifts that alter results.

Also, even though I have degraded the requirement (I would have picked 0.001 for rounding, or I guess 0.0005 actually), the test is still under 4 mm, which is an awful lot to ask in practical terms.  What place lacks waves of that size, or wind-driven variations of that size?

I'll leave this issue open for a while longer.  Let's say 2 days longer.  That will give @bluepost22 and @richardsc a chance to put ideas forward.

**Comment by dankelley** [comment]
@bluepost22 what do you mean by 'There's also a pretty big time delta on alternate low tides.'?  It's confusing at this stage, since I think we are all busy with other things, and there are lots of new issues coming from this one.  Lets ignore the graph we see above, which was before lots of changes.

The present state (just generated now, with the very latest update of the "develop" branch, which we need to be using for *all* tests) is as in the PNG below, when I run the code that is at the official test site (https://github.com/dankelley/oce-issues/blob/main/21xx/2146/2146a.R).  That test site is the place to look, not at any code presented in previous comments.

In the png, the top panel has noaa in black and then oce (offset) in a very thick dashed red curve.  The point of the thickness is to show that they align very well.  The other two panels show that the agreement is within 3mm, for this 800-m tide.  That's not as tight as I would like to see, but if the goal is to get the timing of high and low tides right, I think oce is okay.

The code explains what the middle and bottom panels are.  In words, the middle is what we get when oce picks frequencies, and the bottom is what we get when oce *infers* frequencies, based on NOAA statements about "speed".  I see the middle and bottom panels as being identical by eye.  That means that the very slight difference in frequency is not affecting much.  (The very worst mismatch in freq is for SA, and it has fractional difference of under 1 part in 100,000 parts.)

Long story short -- I think the timing is excellent.  I think you might be referring to previous picture in this thread or some other thread, but the goal is to work with the present oce, with the present test code.  Let's concentrate on that.

If there are new issues on tRef or other things, then they need to be new issues.  Granular issues are important to me, especially now that classes are on.  They are the only way to prevent issues going stagnant, which ends up using a lot of time lost to restarts.

How's this sound, @bluepost22?

![2146a](https://github.com/dankelley/oce/assets/99469/9033aa71-21a4-4db0-b050-97f9b6e41e01)

**Comment by dankelley** [comment]
I've moved discussion of the low-frequency misfit to #2150, in accordance with a desire to avoid topic drift.  We want users to be able to tell whether an issue is relevant to them by scanning titles.

**Comment by dankelley** [comment]
@bluepost22 I altered the docs to mention about reference level. If you click on bb64283895e73bcd454c5d0f54d86df4511f5a48 you can see the difference in R code (and in the Rd, documentation code, that is created by the R code).

I still don't know why my inferred `zoffset` value differs from the MSL value, but I also don't know how to make progress on that topic.  It could be a new issue.  This present issue -- **as defined by its title** has been addressed, and so I am closing it.
